File: app/src/main/java/org/alkaline/taskbrain/dsl/language/IdempotencyAnalyzer.kt
package org.alkaline.taskbrain.dsl.language

/**
 * Analyzes AST expressions for idempotency.
 *
 * Idempotency rules:
 * - Idempotent mutations (allowed at top-level): property assignments (.path, .name)
 * - Non-idempotent mutations (require button/schedule): .append(), new()
 * - Idempotent operations: maybe_new(), find(), calculations, property access
 * - Propagation: A StatementList is non-idempotent if any statement is non-idempotent
 *
 * Non-idempotent directives at top-level will show an error suggesting
 * to use `button` or `schedule` constructs.
 */
object IdempotencyAnalyzer {

    /**
     * Result of idempotency analysis.
     */
    data class AnalysisResult(
        val isIdempotent: Boolean,
        val nonIdempotentReason: String? = null
    ) {
        companion object {
            val IDEMPOTENT = AnalysisResult(true)
            fun nonIdempotent(reason: String) = AnalysisResult(false, reason)
        }
    }

    /**
     * Non-idempotent method names that require explicit triggers.
     */
    private val NON_IDEMPOTENT_METHODS = setOf("append")

    /**
     * Non-idempotent function names that require explicit triggers.
     */
    private val NON_IDEMPOTENT_FUNCTIONS = setOf("new")

    /**
     * Idempotent property names that can be assigned at top-level.
     * These set a value, so repeated execution produces the same result.
     */
    private val IDEMPOTENT_PROPERTIES = setOf("path", "name")

    /**
     * Analyze an expression for idempotency.
     *
     * Milestone 8: Added LambdaExpr support.
     * Phase 0b: Added LambdaInvocation support.
     *
     * @param expr The expression to analyze
     * @return AnalysisResult indicating whether the expression is idempotent
     */
    fun analyze(expr: Expression): AnalysisResult {
        return when (expr) {
            // Literals are always idempotent
            is NumberLiteral -> AnalysisResult.IDEMPOTENT
            is StringLiteral -> AnalysisResult.IDEMPOTENT

            // Variable references are idempotent (just reading a value)
            is VariableRef -> AnalysisResult.IDEMPOTENT

            // Current note reference is idempotent (just a reference)
            is CurrentNoteRef -> AnalysisResult.IDEMPOTENT

            // Property access is idempotent (reading a property)
            is PropertyAccess -> analyze(expr.target)

            // Pattern expressions are idempotent (they're just values)
            is PatternExpr -> AnalysisResult.IDEMPOTENT

            // Lambda expressions - analyze the body for idempotency
            is LambdaExpr -> analyze(expr.body)

            // Lambda invocations - analyze lambda body and arguments
            is LambdaInvocation -> analyzeLambdaInvocation(expr)

            // Function calls - check if the function is non-idempotent
            is CallExpr -> analyzeCallExpr(expr)

            // Method calls - check if the method is non-idempotent
            is MethodCall -> analyzeMethodCall(expr)

            // Assignments - check what's being assigned
            is Assignment -> analyzeAssignment(expr)

            // Statement lists - non-idempotent if any statement is
            is StatementList -> analyzeStatementList(expr)
        }
    }

    private fun analyzeLambdaInvocation(expr: LambdaInvocation): AnalysisResult {
        // Check the lambda body
        val lambdaResult = analyze(expr.lambda.body)
        if (!lambdaResult.isIdempotent) {
            return lambdaResult
        }

        // Check arguments
        for (arg in expr.args) {
            val argResult = analyze(arg)
            if (!argResult.isIdempotent) {
                return argResult
            }
        }
        for (namedArg in expr.namedArgs) {
            val argResult = analyze(namedArg.value)
            if (!argResult.isIdempotent) {
                return argResult
            }
        }

        return AnalysisResult.IDEMPOTENT
    }

    private fun analyzeCallExpr(expr: CallExpr): AnalysisResult {
        // Check if the function itself is non-idempotent
        if (expr.name in NON_IDEMPOTENT_FUNCTIONS) {
            return AnalysisResult.nonIdempotent(
                "${expr.name}() creates new data and requires an explicit trigger. " +
                "Wrap in button() or schedule() to execute."
            )
        }

        // Check arguments for non-idempotency
        for (arg in expr.args) {
            val argResult = analyze(arg)
            if (!argResult.isIdempotent) {
                return argResult
            }
        }
        for (namedArg in expr.namedArgs) {
            val argResult = analyze(namedArg.value)
            if (!argResult.isIdempotent) {
                return argResult
            }
        }

        return AnalysisResult.IDEMPOTENT
    }

    private fun analyzeMethodCall(expr: MethodCall): AnalysisResult {
        // Check if the method is non-idempotent
        if (expr.methodName in NON_IDEMPOTENT_METHODS) {
            return AnalysisResult.nonIdempotent(
                ".${expr.methodName}() modifies data and requires an explicit trigger. " +
                "Wrap in button() or schedule() to execute."
            )
        }

        // Check target and arguments
        val targetResult = analyze(expr.target)
        if (!targetResult.isIdempotent) {
            return targetResult
        }

        for (arg in expr.args) {
            val argResult = analyze(arg)
            if (!argResult.isIdempotent) {
                return argResult
            }
        }
        for (namedArg in expr.namedArgs) {
            val argResult = analyze(namedArg.value)
            if (!argResult.isIdempotent) {
                return argResult
            }
        }

        return AnalysisResult.IDEMPOTENT
    }

    private fun analyzeAssignment(expr: Assignment): AnalysisResult {
        // Property assignments to idempotent properties are allowed
        // Example: [.path: "foo"] or [.name: "bar"]
        if (expr.target is PropertyAccess) {
            val propAccess = expr.target
            if (propAccess.property in IDEMPOTENT_PROPERTIES) {
                // Property assignment is idempotent, but check the value
                return analyze(expr.value)
            }
            // Non-idempotent property assignment (if we add more properties later)
            // For now, only path and name exist and are idempotent
        }

        // Variable assignments - check if the value is idempotent
        // Example: [x: new(path: "foo")] - non-idempotent because of new()
        return analyze(expr.value)
    }

    private fun analyzeStatementList(expr: StatementList): AnalysisResult {
        // A statement list is non-idempotent if ANY statement is non-idempotent
        for (statement in expr.statements) {
            val result = analyze(statement)
            if (!result.isIdempotent) {
                return result
            }
        }
        return AnalysisResult.IDEMPOTENT
    }
}

File: web/src/dsl/language/IdempotencyAnalyzer.ts
import type { Expression } from './Expression'

export interface AnalysisResult {
  isIdempotent: boolean
  nonIdempotentReason: string | null
}

const IDEMPOTENT: AnalysisResult = { isIdempotent: true, nonIdempotentReason: null }

function nonIdempotent(reason: string): AnalysisResult {
  return { isIdempotent: false, nonIdempotentReason: reason }
}

const NON_IDEMPOTENT_METHODS = new Set(['append'])
const NON_IDEMPOTENT_FUNCTIONS = new Set(['new'])
const ACTION_WRAPPER_FUNCTIONS = new Set(['button', 'schedule'])
const IDEMPOTENT_PROPERTIES = new Set(['path', 'name'])

export function analyze(expr: Expression): AnalysisResult {
  switch (expr.kind) {
    case 'NumberLiteral':
    case 'StringLiteral':
    case 'VariableRef':
    case 'CurrentNoteRef':
    case 'PatternExpr':
    case 'OnceExpr':
    case 'RefreshExpr':
      return IDEMPOTENT

    case 'PropertyAccess':
      return analyze(expr.target)

    case 'LambdaExpr':
      return analyze(expr.body)

    case 'LambdaInvocation': {
      const lr = analyze(expr.lambda.body)
      if (!lr.isIdempotent) return lr
      for (const arg of expr.args) {
        const ar = analyze(arg)
        if (!ar.isIdempotent) return ar
      }
      for (const na of expr.namedArgs) {
        const ar = analyze(na.value)
        if (!ar.isIdempotent) return ar
      }
      return IDEMPOTENT
    }

    case 'CallExpr': {
      if (NON_IDEMPOTENT_FUNCTIONS.has(expr.name)) {
        return nonIdempotent(
          `${expr.name}() creates new data and requires an explicit trigger. Wrap in button() or schedule() to execute.`,
        )
      }
      if (ACTION_WRAPPER_FUNCTIONS.has(expr.name)) {
        const firstArg = expr.args[0]
        if (firstArg) {
          const fr = analyze(firstArg)
          if (!fr.isIdempotent) return fr
        }
        return IDEMPOTENT
      }
      for (const arg of expr.args) {
        const ar = analyze(arg)
        if (!ar.isIdempotent) return ar
      }
      for (const na of expr.namedArgs) {
        const ar = analyze(na.value)
        if (!ar.isIdempotent) return ar
      }
      return IDEMPOTENT
    }

    case 'MethodCall': {
      if (NON_IDEMPOTENT_METHODS.has(expr.methodName)) {
        return nonIdempotent(
          `.${expr.methodName}() modifies data and requires an explicit trigger. Wrap in button() or schedule() to execute.`,
        )
      }
      const tr = analyze(expr.target)
      if (!tr.isIdempotent) return tr
      for (const arg of expr.args) {
        const ar = analyze(arg)
        if (!ar.isIdempotent) return ar
      }
      for (const na of expr.namedArgs) {
        const ar = analyze(na.value)
        if (!ar.isIdempotent) return ar
      }
      return IDEMPOTENT
    }

    case 'Assignment': {
      if (expr.target.kind === 'PropertyAccess' && IDEMPOTENT_PROPERTIES.has(expr.target.property)) {
        return analyze(expr.value)
      }
      return analyze(expr.value)
    }

    case 'StatementList':
      for (const stmt of expr.statements) {
        const r = analyze(stmt)
        if (!r.isIdempotent) return r
      }
      return IDEMPOTENT
  }
}

export function checkUnwrappedMutations(expr: Expression): AnalysisResult {
  switch (expr.kind) {
    case 'OnceExpr':
      return IDEMPOTENT

    case 'Assignment': {
      if (expr.target.kind === 'PropertyAccess' && IDEMPOTENT_PROPERTIES.has(expr.target.property)) {
        const targetPath = formatPropertyPath(expr.target)
        return nonIdempotent(
          `Property assignment ${targetPath}: requires once[] wrapper. Use: once[${formatAssignment(expr)}] to prevent unintended re-execution.`,
        )
      }
      return checkUnwrappedMutations(expr.value)
    }

    case 'StatementList':
      for (const stmt of expr.statements) {
        const r = checkUnwrappedMutations(stmt)
        if (!r.isIdempotent) return r
      }
      return IDEMPOTENT

    case 'RefreshExpr':
      return checkUnwrappedMutations(expr.body)

    default:
      return IDEMPOTENT
  }
}

function formatPropertyPath(expr: Expression): string {
  if (expr.kind !== 'PropertyAccess') return '...'
  let prefix: string
  switch (expr.target.kind) {
    case 'CurrentNoteRef': prefix = '.'; break
    case 'PropertyAccess': prefix = formatPropertyPath(expr.target) + '.'; break
    case 'MethodCall': prefix = formatMethodCall(expr.target) + '.'; break
    default: prefix = '...'; break
  }
  return prefix + expr.property
}

function formatMethodCall(expr: Expression): string {
  if (expr.kind !== 'MethodCall') return '...'
  let prefix: string
  switch (expr.target.kind) {
    case 'CurrentNoteRef': prefix = '.'; break
    case 'PropertyAccess': prefix = formatPropertyPath(expr.target) + '.'; break
    case 'MethodCall': prefix = formatMethodCall(expr.target) + '.'; break
    default: prefix = '...'; break
  }
  const args = expr.args.map(formatValue).join(', ')
  return `${prefix}${expr.methodName}(${args})`
}

function formatValue(expr: Expression): string {
  switch (expr.kind) {
    case 'StringLiteral': return `"${expr.value}"`
    case 'NumberLiteral': return String(expr.value)
    default: return '...'
  }
}

function formatAssignment(expr: Expression): string {
  if (expr.kind !== 'Assignment') return '...'
  const target = expr.target.kind === 'PropertyAccess' ? formatPropertyPath(expr.target) : '...'
  const value = formatValue(expr.value)
  return `${target}: ${value}`
}
